### PR TITLE
WiP: Work on refurbishing `publish` (replaces #617)

### PR DESCRIPTION
**Pull request by mih**

### CODE REVIEW

**Comment by yarikoptic** [comment]
Just to keep in mind: May be not for this PR, but I desperately need some way to say "publish --inherit" which, if no sibling is defined would figure out parameters from its super, and call create_publication... and then does the push

**Comment by mih** [comment]
Yes, I had hit that myself yesterday.

**Comment by yarikoptic** [comment]
another one related -- eventually I think we should make `publish` use `annex sync --content` and its `git-annex-preferred-content` specifications, so we could set up expressions (per each repository) on what to be uploaded.  Alternative is to provide per remote `remote.<name>.annex-copy-options` (@joeyh -- I wondered why there is none?) and call copy with it (without requiring any additional option to publish command) if it is defined

Our most common use cases as far as I see it

1. publish all data available locally (pretty much full sync to remote, but sync would also fetch data from remote, which might be undesirable)
2. publish derived data available ONLY locally. Probably could be quite a simple expression BUT special attention should be paid to content from archives, since it is not per se (according to git-annex branch) available from e.g. original location.

**Comment by joeyh** [comment]
Yaroslav Halchenko wrote:
> Alternative is to provide per remote remote.<name>.annex-copy-options 
> (@joeyh -- I wondered why there is none?)

You seem to be referring to some annex.copy-optiopns configuration, but
there is no such thing, so I doh't know what you mean.

-- 
see shy jo

**Comment by yarikoptic** [comment]
On Thu, 15 Sep 2016, Joey Hess wrote:

>    Yaroslav Halchenko wrote:
>    > Alternative is to provide per remote remote.<name>.annex-copy-options
>    > (@joeyh -- I wondered why there is none?)

>    You seem to be referring to some annex.copy-optiopns configuration, but
>    there is no such thing, so I doh't know what you mean.

That is what was my question:  why there is no option for 'annex copy
--to=name' to specify "preferred content" which should be copied
to remote end any time I call 'git annex copy --to=name'.  Now I see
that there is '--auto' and then it would rely on preferred content
setting.  BUT, since preferred content settings are recorded for the
repository within git-annex branch, then there could be no difference in
what is copied with copy --auto depending on originating repository.
Which could be achieved if there was e.g. that git config setting which
would have recorded "preferred content to be copied" for any 'annex
copy --to=name' I run


-- 
Yaroslav O. Halchenko
Center for Open Neuroscience     http://centerforopenneuroscience.org
Dartmouth College, 419 Moore Hall, Hinman Box 6207, Hanover, NH 03755
Phone: +1 (603) 646-9834                       Fax: +1 (603) 646-1419
WWW:   http://www.linkedin.com/in/yarik

**Comment by joeyh** [comment]
Yaroslav Halchenko wrote:
> BUT, since preferred content settings are recorded for the
> repository within git-annex branch, then there could be no difference in
> what is copied with copy --auto depending on originating repository.

Indeed, it depends on the repostory that's being copied to.

> Which could be achieved if there was e.g. that git config setting which
> would have recorded "preferred content to be copied" for any 'annex
> copy --to=name' I run

So you want to have vary what content the local repository prefers to
copy to remotes, independent of the remote's preferred content settings?

Can I have a simple motivating example?

-- 
see shy jo

**Comment by yarikoptic** [comment]
On Thu, 15 Sep 2016, Joey Hess wrote:

>    Yaroslav Halchenko wrote:
>    > BUT, since preferred content settings are recorded for the
>    > repository within git-annex branch, then there could be no difference in
>    > what is copied with copy --auto depending on originating repository.

>    Indeed, it depends on the repostory that's being copied to.

>    > Which could be achieved if there was e.g. that git config setting which
>    > would have recorded "preferred content to be copied" for any 'annex
>    > copy --to=name' I run

>    So you want to have vary what content the local repository prefers to
>    copy to remotes, independent of the remote's preferred content settings?

>    Can I have a simple motivating example?

Thanks for asking .  Let's see

annex A:  provides git-annex without any data.  keys have urls to
    fetch content via datalad-archives special remote which in turn requests
    annex to fetch tarballs, which are then fetched by datalad special
    remote from a remote service X which requires authentication

annex B:  clone of A.  has some content fetched from X from
    datalad-archives, which are in turn fetched via datalad  special remote
    from some urls.  Those data files must not be re-distributed.

    But B also got some additional "derived" data files now.

annex C: local clone of A with some files fetched from B.
     B and C are synced so they know that some files have more than
     one copy  (actually not sure yet if having C plays any role here
     ;-) just a real scenario).  But C also must not share any files
     which originally came from datalad/datalad-archives special remotes


annex D: public repository (we create and publish to)  which must not
    contain any file which was originally obtained from datalad-archives
    which came from datalad (with authentication)

    I would like to copy from B to D those new "derived" data files,
    and only them.


annex E:  clone of D, where I create some additional files and want all
    of the files from E to be copied to D.  Actually some of those might
    be coming from datalad-archives  special remote as well but not from
    "protected" sources.  I just know  that I have nothing on E which 
    cannot be shared, so I want to copy EVERYTHING to D.


so I see B and C needing one policy  and E having  another "policy" on
"what to copy to D", and that is why thought that it could not be
achieved with a single recorded in git-annex "preferred content".

Makes sense?

-- 
Yaroslav O. Halchenko
Center for Open Neuroscience     http://centerforopenneuroscience.org
Dartmouth College, 419 Moore Hall, Hinman Box 6207, Hanover, NH 03755
Phone: +1 (603) 646-9834                       Fax: +1 (603) 646-1419
WWW:   http://www.linkedin.com/in/yarik

**Comment by bpoldrack** [comment]
Hm. At a first glance I don't see why you would need different policies, since the policy for B and C would exclude things, you don't have on E anyway, right? So what would be the harm to use the same policy for E?

**Comment by joeyh** [comment]
Seems like you can do this at the command line level by eg
running in B

	git annex copy --to D --not --in X

Thus avoiding any of the non-public files in X being sent to D.

This same command line would also presumably work when run in E,
since nothing in E is stored on X anyway.

Preferred content expressions don't have an equivilant of --in, but
a similar thing can be accomplished using groups. 

If X is added to group "private", then the preferred content expression
for D would include "not copies=private:1"

I don't see a case yet for having different repositories have their own
different opinions about the preferred content for D. I'm pretty sure
that implementing such a thing would have high foot-gun potential
because it would be easy to end up with a situation where B thinks that
D should not have file F, while E, with a different configuration,
thinks that D does want file F. And so F could be copied and dropped
repeatedly as the repositories fight over their difference of opinion.

-- 
see shy jo

File: datalad/distribution/publish.py [hunk 1]
```diff
@@ -116,16 +112,27 @@ class Publish(Interface):
...
-        annex_copy_opts=Parameter(
-            args=("--annex-copy-opts",),
-            metavar='OPT_STR',  # better name?
-            doc="options passed to 'annex copy'",
-            constraints=EnsureStr() | EnsureNone(),))
+        recursive=recursion_flag,
+        recursion_limit=recursion_limit,
+        git_opts=git_opts,
+        annex_opts=annex_opts,
+        annex_copy_opts=annex_copy_opts
```
Comment (yarikoptic) [inline datalad/distribution/publish.py]: while you are at it -- so how do we publish with all the data? ;)

**Comment by mih** [comment]
I am closing this now. I am not working on it, and it seems I will be stuck with other issues for a long time.

**Pull request closed without merging**